cpp/src/arrow/util/future.cc

```diff
@@ -0,0 +1,300 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/future.h"
+
+#include <algorithm>
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <numeric>
+
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+
+// Shared mutex for all FutureWaiter instances.
+// This simplifies lock management compared to a per-waiter mutex.
+// The locking order is: global waiter mutex, then per-future mutex.
+//
+// It is unlikely that many waiter instances are alive at once, so this
+// should ideally not limit scalability.
+static std::mutex global_waiter_mutex;
```

Comment:
No, it's more delicate. With a per-waiter mutex, `FutureImpl::DoMarkFinishedOrFailed` has to check whether it has a waiter before locking the waiter. But before checking whether it has a waiter, it must take its own lock. And then it's violating the "waiter then future" locking order.

The alternative is to switch the locking order to "future then waiter". It shifts the implementation complexity to the `Waiter` constructor. Currently, the waiter locks itself and then locks every future individually. It would have to first lock all its futures at once, then lock itself. This creates a much larger critical section. Also, theoretically we might also need to lock those futures in deterministic order (I don't think it would matter right now, because a future can only have a single waiter, but I'm not entirely sure).

So the simplest solution, in terms of implementation complexity, really seems to be the global waiter lock. Perhaps at some point we'll notice it creates too much contention, in which case we'll have to switch to a more involved implementation strategy (or to a third-party library that has this all solved efficiently :-)).